### DISCUSSION TITLE
dispatchReactUnityEvent is not a function

### DISCUSSION THREAD

**Discussion by ReginaldBull** [jeffreylanters/react-unity-webgl#440] [Help Wanted!]
### Please avoid duplicates

- [X] I checked [all open issues](https://github.com/jeffreylanters/react-unity-webgl/issues) and [discussions](https://github.com/jeffreylanters/react-unity-webgl/discussions) and none of them matched my problem.

### Language and Compiler

TypeScript's built-in Compiler

### What environment are you using?

Other

### When does your problem occur?

When the Unity App is running

### What does your problem relate to?

I don't know

### React-Unity-WebGL Version

9.0.4

### React Version

17.0.2

### Unity Version

2020.3.37f1

### What happened?

Hello,

Followed the instructions in the documentation. In Unity I created under Plugins/WebGL a file "React.jslib" with following content:

```js
mergeInto(LibraryManager.library, {
    UnityContextInitialized: function (message) {
        try{
            console.log("Unity: " + UTF8ToString(message));
            window.dispatchReactUnityEvent("UnityContextInitialized", UTF8ToString(message));
        } catch(e) {
            console.error(e);
        }
    },
});
```

Inside of Unity

```cs
public sealed class ConfigurationApi : MonoBehaviour, IConfigurationApi
    {
        [DllImport("__Internal")]
        private static extern void UnityContextInitialized(string message);

        public void Initialize(string json)
        {   
            InitializeAsync(json)
                .GetAwaiter()
                .GetResult();
#if UNITY_WEBGL == true && UNITY_EDITOR == false
            UnityContextInitialized("Initialized");
#endif
        }
}
```

The Initialize method is called by the React-Frontend and works.

Now I want to give a message back after the initialization happened. Therefore, I added the UnityContextInitialized("Initialized") call.

While the project is run and opened, the following message shows up in the console output:

`TypeError: window.dispatchReactUnityEvent is not a function.`

I see the log message from this console.log: console.log("Unity: " + UTF8ToString(message));

Any ideas?

### Would you be interested in contributing a fix?

- [ ] yes, I would like to contribute a fix

**Comment by jeffreylanters** [comment]
Hi! Have you registered an event listener. Could you share some more code on your React side of things?

**Comment by ReginaldBull** [comment]
Hello,
Sorry for the long delay!

Currently event listener is registered in the react component:
```typescript
const {
    addEventListener,
    removeEventListener,
    isLoaded,
    sendMessage,
    unityProvider,
    loadingProgression,
  } = useUnityContext({
    loaderUrl: `${TDS_3D_UNITY_URL}/tds-3d-unity-core.loader.js`,
    dataUrl: `${TDS_3D_UNITY_URL}/tds-3d-unity-core.data`,
    frameworkUrl: `${TDS_3D_UNITY_URL}/tds-3d-unity-core.framework.js`,
    codeUrl: `${TDS_3D_UNITY_URL}/tds-3d-unity-core.wasm`,
  });
```

and 

```typscript
/**
   * unity custom event listener logic
   * care! the addEventListener functions are coming from the unityContext
   */
  useEffect(() => {
    addEventListener('error', handleUnityError);
    addEventListener('UnityContextInitialized', handleUnityContextInitialized);
    return () => {
      removeEventListener('error', handleUnityError);
      removeEventListener('UnityContextInitialized', handleUnityContextInitialized);
    };
    // eslint-disable-next-line react-hooks/exhaustive-deps
  }, [addEventListener, removeEventListener]);
```

**Reply by ReginaldBull** [reply]
Any ideas?

**Reply by jeffreylanters** [reply]
Sorry. I've missed your response. 

Can you try opening the browser's inspector and going to the console, then trying to type in `typeof window.dispatchReactUnityEvent`? This should return something like `Function`. By doing this, I can validate whether the bridge is injected correctly. Keep me posted.

**Comment by ReginaldBull** [comment]
Thank you for your response and your patience.
Returns an undefined.

**Reply by jeffreylanters** [reply]
I'm not sure what's going wrong here. Is there any chance I could look into some of your code privately so I can try figure out what's going wrong?

**Reply by ReginaldBull** [reply]
We could arrange something. Code itself is on a private gitlab repository. How can I send you a link privately?

**Reply by jeffreylanters** [reply]
You could send me an invite via the email on my profile. This was I can check out the code and run/debug it locally. But I understand it if that is not something you'd like to do. An other solution would be to send me just some partial React code in order look into.

**Comment by ReginaldBull** [comment]
Hello everyone,

It's been quite a while and I wanted to give an update. Due to some unforeseen side projects, I have been delayed in working on this. This week I tried again, and it worked without any problems.

This means for me that this topic can be closed.

Kind regards

**Reply by zanottipaolo** [reply]
@ReginaldBull Can I ask you how you solved the problem?

**Reply by ReginaldBull** [reply]
@zanottipaolo are you facing the error from above? What versions (Unity, React-Unity-WebGL) are you using?

**Reply by zanottipaolo** [reply]
@ReginaldBull Well not exactly, `window.dispatchReactUnityEvent` returns a function but custom event for communications from Unity to React returns `undefined`. Maybe I should open a new issue about this.

However,
Unity version: 2022.2.6
react-unity-webgl: 9.5.1
